File: sec5/p055.py
n, w = map(int, input().split())
l = list(map(int, input().split()))
dp = [False] * (w+1)
dp[0] = True
for i in range(1, w+1):
    for j in l:
        if i < j:
            continue
        if dp[i-j]:
            dp[i] = True
if dp[w]:
    print('Yes')
else:
    print('No')

print('##################################')

n, w = map(int, input().split())
l = list(map(int, input().split()))
dp = [[False]*(w+1) for _ in range(n+1)]
dp[0][0] = True
for i in range(n):
    for j in range(w+1):
        if dp[i][j]:
            dp[i+1][j] = True
        if j >= l[i] and dp[i+1][j-l[i]]:
            dp[i+1][j] = True
if dp[n][w]:
    print('Yes')
else:
    print('No')

File: sec5/p055.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var scan = bufio.NewScanner(os.Stdin)

const (
	initBufSize = 10000
	maxBufSize = 10000000
)

func init() {
	buf := make([]byte, initBufSize)
	scan.Buffer(buf, maxBufSize)
	scan.Split(bufio.ScanWords)
}

func main() {
	n := NumStdin()
	w := NumStdin()

	a := make([]int, n)
	for i := 0; i < n; i++ {
		a[i] = NumStdin()
	}

	dp := make([]bool, w+1)
	dp[0] = true
	for i := 0; i < n; i++ {
		for j := 0; j <= w; j++ {
			if a[i] <= j {
				dp[j] = dp[j] || dp[j-a[i]]
			}
		}
	}

	if dp[w] {
		fmt.Println("Yes")
	} else {
		fmt.Println("No")
	}
}

func NumStdin() int {
	scan.Scan()
	num, err := strconv.Atoi(scan.Text())
	if err != nil {
		panic(err)
	}
	return num
}
